File: apps/ios/Exponential/UI/GettingStarted/GettingStartedCards.swift
import ExpUI
import ExpCore
import SwiftUI

/// The getting-started checklist, mounted under an empty board or an empty
/// team (EXP-698 r5).
///
/// It is the same block web renders in its sidebar and the IDE renders on its
/// welcome page: a section header with the `{done}/{total}` readout, then one
/// glass card per step carrying a state glyph, the step's own icon, its title,
/// a one-line description and the action pill that starts it. The copy lives
/// in `GettingStartedCopy` and is byte-identical on all four clients.
///
/// An empty board is exactly where a new team lands first, and it used to say
/// nothing but "No issues yet" — the checklist turns that dead end into the
/// list of things worth doing next, and disappears on its own once they are
/// all done.
struct GettingStartedCards: View {
    let progress: GettingStartedProgress
    let accountId: String
    let teamId: String
    /// Opens the create-board sheet on the host (it owns the team resolve and
    /// the sheet's `onCreated` landing).
    let onCreateBoard: () -> Void

    @Environment(AppDependencies.self) private var deps
    @Environment(\.pushRoute) private var pushRoute
    @Environment(\.openURL) private var openURL

    /// Flips for a moment after the server step copies its install command —
    /// the pasteboard is silent otherwise.
    @State private var copiedInstall = false

    var body: some View {
        if !progress.entries.isEmpty && !progress.complete {
            VStack(alignment: .leading, spacing: 8) {
                GlassSectionHeader(GettingStartedCopy.sectionTitle) {
                    Text(GettingStartedCopy.progress(done: progress.done, total: progress.total))
                        .font(.caption)
                        .foregroundStyle(.white.opacity(TextOpacity.tertiary))
                }

                ForEach(Array(progress.entries.enumerated()), id: \.element.id) { index, entry in
                    card(entry, stepNumber: index + 1)
                }
            }
            .accessibilityIdentifier("getting-started")
        }
    }

    @ViewBuilder
    private func card(_ entry: GettingStartedProgress.Entry, stepNumber: Int) -> some View {
        // While the signals are still loading, nothing is done or locked yet:
        // every card reads as a plain numbered step.
        let state: GettingStartedProgress.EntryState = progress.loading ? .available : entry.state
        let locked = state == .locked

        VStack(alignment: .leading, spacing: 6) {
            HStack(spacing: 10) {
                stateGlyph(state: state, stepNumber: stepNumber)

                AppIcon(entryIcon(entry.key), size: AppIcon.Size.medium)
                    .foregroundStyle(.white.opacity(TextOpacity.secondary))

                Text(entry.key.title)
                    .font(.subheadline.weight(.medium))
                    .foregroundStyle(.white)
                    .lineLimit(1)

                Spacer(minLength: 8)
            }

            // Two lines, and the CARD grows to hold them (Android's
            // `maxLines = 2`, web wraps freely). At one line every description
            // and every locked hint ended in an ellipsis a third of the way
            // through its sentence — the shot showed "Runs coding sessions on
            // your machine and registers it as one of y…", which tells a new
            // user nothing. `fixedSize` is what stops the VStack from
            // compressing the label back to one line inside a fixed-height
            // card.
            Text(locked ? lockedHint(entry) : entry.key.entryDescription)
                .font(.caption)
                .foregroundStyle(.white.opacity(TextOpacity.tertiary))
                .lineLimit(2)
                .fixedSize(horizontal: false, vertical: true)
                .frame(maxWidth: .infinity, alignment: .leading)

            if state != .done {
                GlassPill(
                    actionLabel(entry.key),
                    size: .sm,
                    mode: .action { perform(entry.key) },
                    enabled: !locked
                )
            }
        }
        .padding(10)
        .frame(maxWidth: .infinity, alignment: .leading)
        .glassCard()
        .opacity(locked ? 0.6 : 1)
    }

    @ViewBuilder
    private func stateGlyph(
        state: GettingStartedProgress.EntryState,
        stepNumber: Int
    ) -> some View {
        switch state {
        case .done:
            AppIcon(AppIcons.uiSelected, size: AppIcon.Size.medium)
                .foregroundStyle(DesignTokens.Semantic.green)
                .frame(width: 22, height: 22)
        case .locked:
            AppIcon(AppIcons.uiPrivate, size: AppIcon.Size.small)
                .foregroundStyle(.white.opacity(TextOpacity.tertiary))
                .frame(width: 22, height: 22)
        case .available:
            Text("\(stepNumber)")
                .font(.caption.weight(.medium))
                .foregroundStyle(.white.opacity(TextOpacity.tertiary))
                .frame(width: 22, height: 22)
                .overlay(
                    Circle().stroke(GlassTokens.strokeCard, lineWidth: GlassTokens.hairline)
                )
        }
    }

    /// The action pill's label — the copy constant, except while the server
    /// step is flashing its copy confirmation.
    private func actionLabel(_ key: GettingStartedEntryKey) -> String {
        if key == .server && copiedInstall { return "Copied" }
        return key.actionLabel
    }

    /// EXP-548 concepts, not glyphs: the same registry names web and the IDE
    /// resolve, so the four checklists render identical icons.
    private func entryIcon(_ key: GettingStartedEntryKey) -> String {
        switch key {
        case .desktop: AppIcons.uiDevice
        case .github: AppIcons.uiGithub
        case .invite: AppIcons.uiInvite
        case .board: AppIcons.navBoards
        case .coding: AppIcons.navTerminal
        case .action: AppIcons.actionCreate
        case .server: AppIcons.uiServer
        }
    }

    private func lockedHint(_ entry: GettingStartedProgress.Entry) -> String {
        guard let lockedBy = entry.lockedBy else { return entry.key.entryDescription }
        return gettingStartedLockedHint(entry: entry.key, lockedBy: lockedBy)
    }

    private func perform(_ key: GettingStartedEntryKey) {
        switch key {
        case .desktop:
            openURL(AppConstants.desktopReleasesUrl)
        case .github, .invite:
            // Both live in team settings: repositories connect the GitHub App,
            // members mint the invite link.
            pushRoute(.teamSettings(accountId: accountId, teamId: teamId))
        case .board:
            onCreateBoard()
        case .coding:
            pushRoute(.agents)
        case .action:
            pushRoute(.actions)
        case .server:
            copyInstallSnippet()
        }
    }

    private func copyInstallSnippet() {
        let instanceUrl = deps.auth.accounts.first { $0.id == accountId }?.instanceUrl
        guard let origin = WebLinks.normalizedBase(instanceUrl) else { return }
        Platform.copyToPasteboard(AppConstants.serverInstallSnippet(origin: origin))
        copiedInstall = true
        Task {
            try? await Task.sleep(for: .seconds(2))
            copiedInstall = false
        }
    }
}

// MARK: - Environment

/// What a screen needs to render the checklist: the live progress, the team it
/// is scoped to, and the host's create-board entry point (the board step opens
/// the very same `CreateBoardSheet` the empty-team state does).
///
/// `IssuesHomeView` owns it and publishes it into the environment, so the
/// Issues ROOT list renders the checklist under its empty state while a pushed
/// `.board` — a second board on a team that already has one — does not.
/// Like `PushRouteAction`, it owns its progress model and takes its handler
/// after construction, so the host holds ONE instance in `@State` and hands
/// the same object to the environment on every body pass — a fresh box per
/// pass would invalidate every reader of `\.gettingStarted` on every render.
@MainActor
final class GettingStartedContext {
    let progress = GettingStartedProgress()
    /// Opens the host's `CreateBoardSheet`; a no-op until the host wires it.
    var onCreateBoard: () -> Void = {}

    init() {}
}

private struct GettingStartedContextKey: EnvironmentKey {
    static let defaultValue: GettingStartedContext? = nil
}

extension EnvironmentValues {
    /// nil on every surface that does not host the checklist.
    var gettingStarted: GettingStartedContext? {
        get { self[GettingStartedContextKey.self] }
        set { self[GettingStartedContextKey.self] = newValue }
    }
}

File: apps/android/app/src/main/java/com/exponential/app/ui/gettingstarted/GettingStartedCards.kt
package com.exponential.app.ui.gettingstarted

import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import com.exponential.app.ui.components.GlassPill
import com.exponential.app.ui.components.PillSize
import com.exponential.app.ui.components.SectionHeader
import com.exponential.app.ui.icons.ExpIcons
import com.exponential.app.ui.theme.DesignTokens
import com.exponential.app.ui.theme.GlassTokens
import com.exponential.app.ui.theme.TextEmphasis
import com.exponential.app.ui.theme.glassCard

/**
 * "What to do next", under an empty board or an empty team (EXP-698 r5) — the
 * mobile twin of web's `GettingStartedCards` and the IDE's inline cards, with
 * the same entries, order, copy and progress count.
 *
 * An empty screen is where a checklist belongs: the two places it mounts are
 * exactly the two where the app has nothing else to show, and it disappears on
 * its own once every entry is done.
 */
@Composable
fun GettingStartedCards(
    state: GettingStartedState,
    onAction: (GettingStartedEntryKey) -> Unit,
    modifier: Modifier = Modifier,
) {
    if (state.entries.isEmpty() || state.complete) return
    Column(
        modifier = modifier.fillMaxWidth().testTag("getting-started"),
        verticalArrangement = Arrangement.spacedBy(8.dp),
    ) {
        SectionHeader(
            GettingStartedCopy.SECTION_TITLE,
            trailing = {
                Text(
                    "${state.done}/${state.total} done",
                    style = MaterialTheme.typography.labelSmall,
                    color = MaterialTheme.colorScheme.onSurface.copy(alpha = TextEmphasis.Tertiary),
                )
            },
        )
        state.entries.forEachIndexed { index, entry ->
            EntryCard(
                entry = entry,
                stepNumber = index + 1,
                loading = state.loading,
                onAction = { onAction(entry.key) },
            )
        }
    }
}

@Composable
private fun EntryCard(
    entry: GettingStartedEntry,
    stepNumber: Int,
    /** Signals still resolving: the entry reads as NEUTRAL (web parity) — no
     *  action offered on a state that may be about to flip to done. */
    loading: Boolean,
    onAction: () -> Unit,
) {
    val locked = entry.state == GettingStartedEntryState.Locked
    Column(
        modifier = Modifier
            .fillMaxWidth()
            // A locked step is still worth reading — it says what to do first
            // — so it dims rather than disappears (web parity).
            .alpha(if (locked) 0.6f else 1f)
            .glassCard()
            .padding(12.dp),
    ) {
        Row(verticalAlignment = Alignment.CenterVertically) {
            StateGlyph(state = entry.state, stepNumber = stepNumber)
            Spacer(Modifier.width(10.dp))
            Icon(
                entryIcon(entry.key),
                contentDescription = null,
                modifier = Modifier.size(16.dp),
                tint = MaterialTheme.colorScheme.onSurface.copy(alpha = TextEmphasis.Secondary),
            )
            Spacer(Modifier.width(8.dp))
            Text(
                GettingStartedCopy.title(entry.key),
                style = MaterialTheme.typography.bodyMedium.copy(fontWeight = FontWeight.Medium),
                color = MaterialTheme.colorScheme.onSurface,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis,
                modifier = Modifier.weight(1f),
            )
        }
        Spacer(Modifier.height(4.dp))
        Text(
            if (locked && entry.lockedBy != null) {
                GettingStartedCopy.lockedHint(entry.key, entry.lockedBy)
            } else {
                GettingStartedCopy.description(entry.key)
            },
            style = MaterialTheme.typography.bodySmall,
            color = MaterialTheme.colorScheme.onSurface.copy(alpha = TextEmphasis.Tertiary),
            // Descriptions and locked hints alike are ONE SENTENCE by
            // contract and wrap on web — cut to a single line they ended
            // mid-word behind an ellipsis. Two lines fits every string in the
            // copy table at this width; the overflow is a guard, not a design.
            maxLines = 2,
            overflow = TextOverflow.Ellipsis,
        )
        // A locked step has nothing to tap yet, and a done one nothing left to
        // do — the pill is for the step you can actually take.
        if (entry.state == GettingStartedEntryState.Available && !loading) {
            Spacer(Modifier.height(10.dp))
            GlassPill(
                GettingStartedCopy.action(entry.key),
                size = PillSize.Sm,
                onClick = onAction,
            )
        }
    }
}

/** Done / locked / the step's number — the checklist's own left column. */
@Composable
private fun StateGlyph(state: GettingStartedEntryState, stepNumber: Int) {
    when (state) {
        GettingStartedEntryState.Done -> Icon(
            ExpIcons.uiSelected,
            contentDescription = null,
            modifier = Modifier.size(GlyphSize),
            tint = DesignTokens.Semantic.Green,
        )
        GettingStartedEntryState.Locked -> Icon(
            ExpIcons.uiPrivate,
            contentDescription = null,
            modifier = Modifier.size(GlyphSize),
            tint = MaterialTheme.colorScheme.onSurface.copy(alpha = TextEmphasis.Tertiary),
        )
        GettingStartedEntryState.Available -> Box(
            modifier = Modifier
                .size(GlyphSize)
                .border(GlassTokens.Hairline, GlassTokens.StrokeCard, CircleShape),
            contentAlignment = Alignment.Center,
        ) {
            Text(
                stepNumber.toString(),
                style = MaterialTheme.typography.labelSmall,
                color = MaterialTheme.colorScheme.onSurface.copy(alpha = TextEmphasis.Tertiary),
            )
        }
    }
}

private val GlyphSize = 22.dp

private fun entryIcon(key: GettingStartedEntryKey): ImageVector = when (key) {
    GettingStartedEntryKey.Desktop -> ExpIcons.uiDevice
    GettingStartedEntryKey.Github -> ExpIcons.uiGithub
    GettingStartedEntryKey.Invite -> ExpIcons.uiInvite
    GettingStartedEntryKey.Board -> ExpIcons.navBoards
    GettingStartedEntryKey.Coding -> ExpIcons.navTerminal
    GettingStartedEntryKey.Action -> ExpIcons.actionCreate
    GettingStartedEntryKey.Server -> ExpIcons.uiServer
}
